File: client/reldat-client.py
#!/usr/bin/python

import re
import socket
import sys

import reldat

def teardown_connection( reldat_conn ):
    # TODO tear down connection
    pass

def transform( reldat_conn, f ):
    # TODO send file to server
    pass

def command_loop( reldat_conn ):
    disconnect = False

    while not disconnect:
        print '>',
        client_input  = raw_input()
        command_parts = re.match( '(\w+)\s*(.+)?', client_input )

        if command_parts is not None:
            command = command_parts.group( 1 )

            if command == 'disconnect':
                disconnect = True
            elif command == 'transform':
                file = command_parts.group( 2 )

                if file is None:
                    print '  Usage: transform <file>'
                else:
                    transform( reldat_conn, file )
            else:
                print '  Unrecognized command. Valid commands are:\n    disconnect\n    transform'

def usage():
    print 'Usage: ./reldata-client.py <host IP address>:<host port> <max receive window size in packets>'
    sys.exit( 0 )

def main( argv ):
    if len( argv ) != 2:
        usage()

    host_machine = re.match( '^(\d{1,3}[.]\d{1,3}[.]\d{1,3}[.]\d{1,3}):(\d{1,5})$', argv[0] )

    if host_machine is None:
        usage()

    ip_address              = host_machine.group( 1 )
    port                    = int( host_machine.group( 2 ) )
    max_receive_window_size = int( argv[1] )

    if port > 65535:
        usage()

    reldat_conn = reldat.Reldat( max_receive_window_size )
    reldat_conn.connect( ip_address, port )
    command_loop( reldat_conn )
    reldat_conn.disconnect()

if __name__ == "__main__":
    main( sys.argv[1:] )

sys.exit( 0 )

File: client/src/ReldatClient.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Scanner;
import java.util.concurrent.Semaphore;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import reldat.ReldatConnection;

public class ReldatClient {
	public static Semaphore mutex = new Semaphore(1);

	public static class CommandReader implements Runnable {
		Scanner scanner;
		String cmd;
		
		public CommandReader() {
			scanner = new Scanner( System.in );
			scanner.useDelimiter( "\n" );
			cmd = "";
		}

		@Override
		public void run() {
			System.out.print( "> " );

			if(scanner.hasNext()) {
				try {
					mutex.acquire();
				} catch (InterruptedException e) {
					e.printStackTrace();
				}
				cmd = scanner.next();
				mutex.release();
			}
		}
		
		public String getCommand()
		{
			String retval = null;

			try {
				mutex.acquire();
				retval = cmd;
				cmd = "";
				mutex.release();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
			
			return retval;
		}
		
		public void closeScanner()
		{
			scanner.close();
		}
	}

	public static void main( String[] args ) throws IOException {
		if( args.length != 2 )
			usage();

		Pattern hostRegex = Pattern.compile( "^(\\d{1,3}[.]\\d{1,3}[.]\\d{1,3}[.]\\d{1,3}):(\\d{1,5})$" );
		Matcher hostMatch = hostRegex.matcher( args[0] );

		if( !hostMatch.matches() )
			usage();

		String ipAddress = hostMatch.group( 1 );
		int port         = Integer.parseInt( hostMatch.group( 2 ) );

		if( port > 65535 )
			usage();

		int maxReceiveWindowSize = Integer.parseInt( args[1] );

		ReldatConnection reldatConn = new ReldatConnection( maxReceiveWindowSize );
		reldatConn.connect( ipAddress, port );
		commandLoop( reldatConn );
        reldatConn.disconnect();
	}

	public static void usage() {
		System.out.println( "Usage: java ReldatClient <host IP address>:<host port> <max receive window size in packets>" );
		System.exit( 0 );
	}

	public static void commandLoop( ReldatConnection reldatConn ) throws IOException {
		boolean disconnect = false;

		CommandReader cr = new CommandReader();
		Thread cmdInput  = new Thread(cr);

		cmdInput.start();

		while( !disconnect ) {
			String clientInput = cr.getCommand();
			
			Pattern commandRegex = Pattern.compile( "(\\w+)\\s*(.+)?" );
			Matcher commandMatch = commandRegex.matcher( clientInput );

			if( commandMatch.matches() )
			{
				String command = commandMatch.group( 1 );

				switch( command )
				{
					case "disconnect":
						disconnect = true;
						return;
					case "transform":
						//System.out.println("Working Directory = " + System.getProperty("user.dir"));
						String fileName = "./client/src/test_file.txt";
						String messageToSend = readFileToString(fileName); 
						System.out.println("TRANSFORMED DATA: " + reldatConn.conversation(messageToSend));
						break;
					default:
						System.out.println( "Unrecognized command. Valid commands are:\n    disconnect\n    transform" );
						break;
				}

				cmdInput.run();
			}
			
			reldatConn.listen();
		}

		try {
			cmdInput.join();
			cr.closeScanner();
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
	}

	public static void transform( ReldatConnection reldatConn, String filename ) {
        try {
			FileReader fr = new FileReader(filename);
            BufferedReader br = new BufferedReader(new FileReader(filename));

			String sCurrentLine = null;
            String newStr = "";

			while ((sCurrentLine = br.readLine()) != null) {
                newStr += sCurrentLine.toUpperCase();
			}

            System.out.println(newStr);
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
	
	public static String readFileToString(String filename) {
		String newStr = "";
        try {
			FileReader fr = new FileReader(filename);
            BufferedReader br = new BufferedReader(new FileReader(filename));

			String sCurrentLine = null;
            
			while ((sCurrentLine = br.readLine()) != null) {
                newStr += sCurrentLine;
			}
            System.out.println(newStr);
		} catch (IOException e) {
			e.printStackTrace();
		}
        return newStr;
	}
	
	
}
